### PR TITLE
Add zombie-cleanup module to garbage-collect stuck ARC runner pods

### PR DESCRIPTION
**Pull request by jeanschmidt**
**Impact:** ARC runner clusters (arc-cbr-production, upstream staging/production)
**Risk:** low

## What
Adds a new `zombie-cleanup` OSDC module that runs as a Kubernetes CronJob every hour, detecting and deleting runner pods that are stuck Pending or Running beyond configurable age thresholds. 

It removes pending pods older than 12h and running pods older than 24h. This prevents zombie runner pods to start when the github job have long been canceled and prevents workflow pods to stick wasting resources when the runner pod failed and didn't clean them up properly.

## Why
ARC runner pods can become zombies — stuck in Pending (e.g. unschedulable due to resource pressure or node failures) or workflow pods can stay Running indefinitely (e.g. a job that never terminates). These orphaned pods waste node capacity, block Karpenter from scaling down nodes, and can accumulate over time. There was no automated mechanism to clean them up.

## How
- The cleanup script uses `lightkube` to list all pods in the `arc-runners` namespace, filtering by age and phase
- Controller-managed pods (owned by ReplicaSet, DaemonSet, StatefulSet, Job) are explicitly skipped so long-lived infrastructure like ARC listener pods and hooks-warmer DaemonSets are never touched
- Pods already being terminated (have `deletionTimestamp`) are also skipped to avoid redundant delete calls
- Pods in `Unknown` phase (typically from node failures) use the running threshold
- 404 on delete is treated as success (pod may have been cleaned up between list and delete)
- Content-addressed image tagging: the deploy script hashes source files to compute the image tag, skipping rebuild when nothing changed
- Image is built locally and pushed to Harbor (not ECR), consistent with how other in-cluster tools are deployed

## Changes
- **`modules/zombie-cleanup/scripts/python/zombie_cleanup.py`** — Python script: lists pods, identifies zombies by phase+age, deletes them. Configurable via env vars (`TARGET_NAMESPACE`, `PENDING_MAX_AGE_HOURS`, `RUNNING_MAX_AGE_HOURS`, `DRY_RUN`)
- **`modules/zombie-cleanup/scripts/python/test_zombie_cleanup.py`** — 431-line test suite covering all helper functions, edge cases (naive timestamps, missing timestamps, 404 races, partial failures), and the `main()` entrypoint
- **`modules/zombie-cleanup/kubernetes/cronjob.yaml`** — CronJob running hourly on base-infra nodes with hardened security context (non-root, read-only rootfs, all caps dropped, seccomp)
- **`modules/zombie-cleanup/kubernetes/rbac.yaml`** — Namespace-scoped ServiceAccount + Role + RoleBinding granting only `get`, `list`, `delete` on pods in `arc-runners`
- **`modules/zombie-cleanup/docker/Dockerfile`** — Minimal Alpine-based image using uv for dependency installation
- **`modules/zombie-cleanup/docker/pyproject.toml`** — Single dependency: `lightkube==0.15.5`
- **`modules/zombie-cleanup/deploy.sh`** — Build script: checks enabled flag, computes content-hash tag, builds/pushes to Harbor, applies RBAC + CronJob with sed-substituted config values
- **`clusters.yaml`** — Adds `zombie-cleanup` to module list for `arc-cbr-production`; adds default config (`pending_max_age_hours: 24`, `running_max_age_hours: 12`, `dry_run: false`)

## Notes
- The module is enabled by default in `clusters.yaml` defaults with `dry_run: false`. Set `zombie_cleanup.dry_run: true` per-cluster to observe before enabling deletion.
- The CronJob has `backoffLimit: 0` and `activeDeadlineSeconds: 300` — a single failed run won't retry and won't block the next hourly execution.
- RBAC is namespace-scoped to `arc-runners` only — the cleanup pod cannot touch pods in other namespaces.

### CODE REVIEW

File: osdc/modules/zombie-cleanup/deploy.sh [hunk 1]
```diff
@@ -0,0 +1,143 @@
+#!/usr/bin/env bash
```
Comment (malfet) [inline osdc/modules/zombie-cleanup/deploy.sh:1]: I don't know much about Kubernetes, but this file looks like something that could be avoided or replaced with 1 liner

File: osdc/modules/zombie-cleanup/scripts/python/zombie_cleanup.py [hunk 2]
```diff
@@ -0,0 +1,179 @@
...
+        config["pending_max_hours"],
+        config["running_max_hours"],
+        config["dry_run"],
+    )
+
+    try:
+        client = Client()
+    except Exception:
+        log.exception("Failed to create Kubernetes client")
+        return 1
```
Comment (malfet) [inline osdc/modules/zombie-cleanup/scripts/python/zombie_cleanup.py]: Any reason why one is better than the other?
```suggestion
    client = Client()
```

File: osdc/modules/zombie-cleanup/scripts/python/zombie_cleanup.py [hunk 3]
```diff
@@ -0,0 +1,179 @@
...
+            log.info("No zombie pods found")
+            return 0
+
+        log.info("Found %d zombie pod(s)", len(zombies))
+        deleted, failed = delete_zombies(client, zombies, config)
+        log.info("Cleanup complete: %d deleted, %d failed", deleted, failed)
+        return 1 if failed > 0 else 0
+    except Exception:
+        log.exception("Cleanup failed")
+        return 1
```
Comment (malfet) [inline osdc/modules/zombie-cleanup/scripts/python/zombie_cleanup.py]: This would make debugging really hard, please remove try/catch block

**Review by malfet — APPROVED** [review]

File: osdc/modules/zombie-cleanup/scripts/python/zombie_cleanup.py [hunk 4]
```diff
@@ -0,0 +1,179 @@
...
+    )
+
+    try:
+        client = Client()
+    except Exception:
+        log.exception("Failed to create Kubernetes client")
+        return 1
+
+    try:
+        zombies = find_zombie_pods(client, config)
```
Comment (huydhn) [inline osdc/modules/zombie-cleanup/scripts/python/zombie_cleanup.py:176]: As this module could kill pods by their age, do you want a safety valve here to see how many zombies it finds or the % of zombie pods and check if that number is sane before actually do the deleting?

My thought process here is that if an abnormal number is captured, i.e. 50% of the pods are zombies, do you trust the module to automatically kill half of the pods?

File: osdc/modules/zombie-cleanup/scripts/python/zombie_cleanup.py [hunk 5]
```diff
@@ -0,0 +1,179 @@
...
+        if not zombies:
+            log.info("No zombie pods found")
+            return 0
+
+        log.info("Found %d zombie pod(s)", len(zombies))
```
Comment (huydhn) [inline osdc/modules/zombie-cleanup/scripts/python/zombie_cleanup.py]: Are these logs going to Grafana? I assume yes, so we can get the logs there to debug issues? For example, to check if a pod is killed as a zombie or exit normally

**Review by huydhn — APPROVED** [review]

File: osdc/modules/zombie-cleanup/deploy.sh (hunk 1 above)
Comment (jeanschmidt) [reply osdc/modules/zombie-cleanup/deploy.sh:1]: Yeahp, this is a explicit technical debt.

I dediced to not go with plumi right away and coordinate deployments with just. So, when we need to do some templating it goes to bash or python scripts. Mostly I try to rely on Python, but a few simple things end up just in bash.

Let me land this one and I'll follow up with replacing them with python, this will be a step stone to move to plumi and have a simpler release script/process

File: osdc/modules/zombie-cleanup/scripts/python/zombie_cleanup.py (hunk 3 above)
Comment (jeanschmidt) [reply osdc/modules/zombie-cleanup/scripts/python/zombie_cleanup.py]: excelent finding

File: osdc/modules/zombie-cleanup/scripts/python/zombie_cleanup.py (hunk 5 above)
Comment (jeanschmidt) [reply osdc/modules/zombie-cleanup/scripts/python/zombie_cleanup.py]: no, let me add this.

File: osdc/modules/zombie-cleanup/scripts/python/zombie_cleanup.py (hunk 4 above)
Comment (jeanschmidt) [reply osdc/modules/zombie-cleanup/scripts/python/zombie_cleanup.py:176]: Good point, we can put some safety guardrails and alert in case those are violated. At worst intervening manually would force us to be aware if something is causing excesive leackages

**Comment by jeanschmidt** [comment]
the change became much more significant due to adding monitoring
